GliaWidgets/ViewController/Chat/ChatViewController.swift

```diff
@@ -48,31 +48,54 @@ class ChatViewController: EngagementViewController, MediaUpgradePresenter,
     }
 
     // swiftlint:disable function_body_length
-    private func bind(viewModel: ChatViewModel, to view: ChatView) {
+    private func bind(to view: ChatView) {
```

Comment:
The thing with this solution as i also told you in DMs, is that, while it works, it also means you have to set the view controller's view model before calling this. It's not a contained function, it depends on other factors. Also the name of it loses its meaning, because what is being binded to the view? The previous version was contained and it showed its purpose through the name alone. If others could recommend a way in which we still fix the leak but we keep the previous method signature, it would be ideal. I read that one can capture a parameter weakly, with something like

```
private func bind(viewModel: ChatViewModel, to view: ChatView) {
    weak var weakViewModel = viewModel

    view.numberOfSections = {
        weakViewModel.numberOfSections
    }
    ...
```

What do you guys think?

Reply:
Another option would be to set the view controller's viewModel at first, like it's being done lower with call. That way we don't have to set the view controller's view model before every execution of `bind`. Although that would mean the function has side effects that maybe are not explicit in the name.

Reply:
1. I tried out the suggestion you outlined by declaring weakViewModel. Unfortunately, it brought back the memory leak and none of the mentioned objects were deallocated when the engagement ended.
2. I agree that having the name Bind on the method doesn't give any context anymore. What would be a better name for it in the current scenario?
3. I did catch that last comment. Could you please elaborate on that a bit?

Reply:
`private func bind(viewModel: ChatViewModel, to view: ChatView) {
        view.header.showBackButton()
        view.header.showCloseButton()

        view.numberOfSections = { [weak viewModel] in
            viewModel?.numberOfSections
        }
        view.numberOfRows = { [weak viewModel] rows in
            viewModel?.numberOfItems(in: rows)
`

This however does work.

Reply:
`private func bind(viewModel: ChatViewModel, to view: ChatView) {
        view.header.showBackButton()
        view.header.showCloseButton()

        view.numberOfSections = { [weak viewModel] in
            viewModel?.numberOfSections
        }
        view.numberOfRows = { [weak viewModel] rows in
            viewModel?.numberOfItems(in: rows)`

This however works

Reply:
![Screenshot 2022-11-16 at 13 19 33](https://user-images.githubusercontent.com/111733845/202167094-097d76b1-09a7-4143-a044-1a7f776a781f.png)

This however works.

Reply:
For 3, I meant that something that we can also do is

```
private func bind(viewModel: ..., to view: ...) {
    self.viewModel = viewModel
    ...
```

That way at least the dependency that this function has on `self.viewModel` is very explicit, and it cannot happen that we call bind and we forget to set `self.viewModel` first. Same thing happens in `CallViewModel.init`, where you initially set `self.call = call` and then just use `self.call` from then onwards.

If my suggestion didn't work, then I guess that it could just be called `bindViewModel(to view:)`. But maybe someone that knows more about memory management than me could suggest if something similar to my original solution would work, because I think that that would be the best way to fix it. To send the view model instead of depending on `self.viewModel`

Reply:
`[weak viewModel]` is also nice. That would be good because we can still use `bind(viewModel: ..., to view:...)` which is a nice name, and we don't depend on `self.viewModel` anymore. What do others think?